### DISCUSSION TITLE
Side by side maps

### DISCUSSION THREAD

**Discussion by mlandsfeld** [tethysplatform/tethys#672] [Q&A]
I am trying to get two GeoServer maps to display side by side on my app home page but I am only able to see one map at a time. I am using the app_two_column.html file as my home page and I have inserted a gizmo tags in the each column and I can send the map_options to both of them but only one will display. The map_wrapper div looks fine in both but one of the maps divs will not contain the canvas tag with class=ol.viewport. I can send the map options to either one and they work fine but when I send both only one is activated. Is there a setting for this? My home.html is listed below. Thanks!

{% extends "tethys_apps/app_header_content.html" %}

{% load static tethys_gizmos %}
{% load tethys_gizmos %}

{% block inner_app_content %}
  <div id="inner-app-content" class="container-fluid">
    <div class="row" style="height:100%;">
    
	    <div class="col-md-6" style="max-height:100%;overflow: auto;">
		    {# left column content #}
	        {% block app_content_lc %}
	          <form method="post">
                {% csrf_token %}
                {% gizmo select_input eo_select_options %}
                <input name="eo_submit" type="submit" value="Update" class="btn btn-default">
              </form>
              {% gizmo eo_map_options %}
            {% endblock %}
	    </div>
	    
	    <div class="col-md-6" style="max-height:100%;overflow: auto;">
		  {# right column content #}
	      {% block app_content_rc %}
	          <form method="post">
                {% csrf_token %}
                {% gizmo select_input forecast_select_options %}
                <input name="forecast_submit" type="submit" value="Update" class="btn btn-default">
              </form>
              {% gizmo forecast_map_options %}
	      
	      {% endblock %}
	    </div>
	    
    </div>
  </div>
{% endblock %}

**Comment by msouff** [comment]
Hi,

The problem seems to be that the map_view gizmo assigns a static ID to the div that contains the map and a few other elements in the map, and since you have two maps, it is assigning the same ID to both of them!

If you inspect your page you'll see something like this:

![image](https://user-images.githubusercontent.com/17328135/111252356-8c672e80-85d6-11eb-9341-efd2546acbf2.png)

IDs are supposed to be unique, that's why one of the maps doesn't work properly.

The map_view gizmo was not created with the idea of having more than one map in the same page. Additional app templates were introduced recently, and the map_view gizmo still needs to be updated accordingly. @swainn can you make an issue out of this so we can look into it?

In the meantime, you can get around this by creating your second map directly (that is, without the help of a gizmo). Here is some basic documentation for that: https://openlayers.org/en/latest/examples/simple.html.

Hopefully that helps.

**Comment by swainn** [comment]
Thanks @msouff. I have created issue #674 as requested. Unfortunately, adding this capability to the MapView gizmo will be a major undertaking since it was initially designed to support only one map on a page. I don't know that it will get addressed in the near future, but it is now on the "wish list". 😄

**Reply by mlandsfeld** [reply]
I was able to get two maps on my page using openlayers directly in javascript. Thanks! However, I thought I would be able to pass the geoserver layer back to the html in the context of the controller when it is rendered. That does not seem to be working as planned. I am sending back a geoserver layer name like this:

'forecast_layer': 'forecast_app:yield_obs_2021_02'

in the context hoping that it would replace the layer in the js:

	  new ol.layer.Image({
		source: new ol.source.ImageWMS({
		  url: 'https://chc-ewx2.chc.ucsb.edu:8443/geoserver/wms',
		  params: {'LAYERS': forecast_layer },
		  serverType: 'geoserver',
		}),
	  }), 

Is there another way to make this happen? 

Also, I tried moving the javascript to main.js and use it as the src in a script tag but the html does not seem to be finding it. I tried this and many other permutations without success:

 <script src='geoserver_app/public/js/main.js'> </script>

Thanks!

**Reply by swainn** [reply]
Hi @mlandsfeld, I think I understand what you are asking: you'd like to know how to access context variables in JavaScript? 

The context variables are not made accessible to the JavaScript automatically. I like to use the "data attribute" pattern to do this. It works something like this:

1. Include the variable in your context object in the controller:

```python
def my_controller(request):
    ...
    context = {
        'forecast_layer': 'forecast_app:yield_obs_2021_02',
    }
    return render(request, 'some_template.html', context)
```

2. Create an element in your HTML template for storing attributes you'd like to access in JavaScript. Be sure to give it an ID so it is easy to find in the JavaScript. Add an attribute for each context variable you'd like access to in the javascript using the pattern `data-<variable-name>`:

```django
<div id="my-attributes" data-forecast-layer="{{ forecast_layer }}"></div>
```

3. In your JavaScript, select the element by its ID and retrieve the data attribute values. JQuery (`$`) makes this very easy:

```javascript
// Note: when using the $.data method, leave off the "data" part of the attribute name
var forecast_layer = $('#my-attributes').data('forecast-layer');
```

**Reply by mlandsfeld** [reply]
That worked nicely, thanks!

Any pointers on how to get a legend on the maps? I am not seeing much in the openlayers documentation on this. I have PNGs that I was using in the map gizmo.
